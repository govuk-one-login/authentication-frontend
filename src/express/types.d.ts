import 'express-serve-static-core';

declare global {
  namespace Express {
    interface Locals {
      sessionId: string;
      strategicAppChannel: boolean;
      webChannel: boolean;
      genericAppChannel: boolean;
      isApp: boolean
    }
    interface Response {
      locals: Locals;
    }
  }
}
