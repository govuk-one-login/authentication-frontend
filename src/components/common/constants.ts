import { PATH_NAMES } from "../../app.constants";
import { getNextState } from "./state-machine/state-machine";

export const SECURITY_CODE_ERROR = "actionType";

export enum SecurityCodeErrorType {
  MfaMaxCodesSent = "mfaMaxCodesSent",
  MfaBlocked = "mfaBlocked",
  MfaMaxRetries = "mfaMaxRetries",
  OtpMaxCodesSent = "otpMaxCodesSent",
  OtpBlocked = "otpBlocked",
  OtpMaxRetries = "otpMaxRetries",
  EmailMaxCodesSent = "emailMaxCodesSent",
  EmailBlocked = "emailBlocked",
  EmailMaxRetries = "emailMaxRetries",
  AuthAppMfaNoCodeValidator = "authAppMfaNoCodeValidator",
  AuthAppMfaMaxRetries = "authAppMfaMaxRetries",
  ChangeSecurityCodesEmailMaxCodesSent = "changeSecurityCodesEmailMaxCodesSent",
  ChangeSecurityCodesEmailBlocked = "changeSecurityCodesEmailBlocked",
  ChangeSecurityCodesEmailMaxRetries = "changeSecurityCodesEmailMaxRetries",
  InvalidPasswordResetCodeMaxRetries = "invalidPasswordResetCodeMaxRetries",
}

export const ERROR_CODES = {
  SESSION_ID_MISSING_OR_INVALID: 1000,
  RESET_PASSWORD_LINK_MAX_RETRIES_REACHED: 1022,
  RESET_PASSWORD_LINK_BLOCKED: 1023,
  NEW_PASSWORD_SAME_AS_EXISTING: 1024,
  ENTERED_INVALID_MFA_MAX_TIMES: 1027,
  VERIFY_EMAIL_CODE_REQUEST_BLOCKED: 1029,
  VERIFY_PHONE_NUMBER_CODE_REQUEST_BLOCKED: 1030,
  VERIFY_PHONE_NUMBER_MAX_CODES_SENT: 1032,
  VERIFY_EMAIL_MAX_CODES_SENT: 1031,
  ENTERED_INVALID_VERIFY_EMAIL_CODE_MAX_TIMES: 1033,
  ENTERED_INVALID_VERIFY_PHONE_NUMBER_CODE_MAX_TIMES: 1034,
  MFA_SMS_MAX_CODES_SENT: 1025,
  MFA_CODE_REQUESTS_BLOCKED: 1026,
  INVALID_MFA_CODE: 1035,
  INVALID_VERIFY_EMAIL_CODE: 1036,
  INVALID_VERIFY_PHONE_NUMBER_CODE: 1037,
  ENTERED_INVALID_PASSWORD_RESET_CODE_MAX_TIMES: 1039,
  INVALID_PASSWORD_MAX_ATTEMPTS_REACHED: 1028,
  RESET_PASSWORD_INVALID_CODE: 1021,
  PASSWORD_IS_COMMON: 1040,
  AUTH_APP_INVALID_CODE_MAX_ATTEMPTS_REACHED: 1042,
  AUTH_APP_INVALID_CODE: 1043,
  ACCOUNT_LOCKED: 1045,
  VERIFY_CHANGE_HOW_GET_SECURITY_CODES_MAX_CODES_SENT: 1046,
  VERIFY_CHANGE_HOW_GET_SECURITY_CODES_CODE_REQUEST_BLOCKED: 1047,
  VERIFY_CHANGE_HOW_GET_SECURITY_CODES_INVALID_CODE: 1048,
};

export const ERROR_CODE_MAPPING: { [p: string]: string } = {
  [ERROR_CODES.ACCOUNT_LOCKED]: pathWithQueryParam(
    PATH_NAMES["ACCOUNT_LOCKED"]
  ),
  [ERROR_CODES.INVALID_PASSWORD_MAX_ATTEMPTS_REACHED]: pathWithQueryParam(
    PATH_NAMES["ACCOUNT_LOCKED"]
  ),
  [ERROR_CODES.MFA_SMS_MAX_CODES_SENT]: pathWithQueryParam(
    PATH_NAMES["SECURITY_CODE_REQUEST_EXCEEDED"],
    SECURITY_CODE_ERROR,
    SecurityCodeErrorType.MfaMaxCodesSent
  ),
  [ERROR_CODES.MFA_CODE_REQUESTS_BLOCKED]: pathWithQueryParam(
    PATH_NAMES["SECURITY_CODE_WAIT"],
    SECURITY_CODE_ERROR,
    SecurityCodeErrorType.MfaBlocked
  ),
  [ERROR_CODES.ENTERED_INVALID_MFA_MAX_TIMES]: pathWithQueryParam(
    PATH_NAMES["SECURITY_CODE_INVALID"],
    SECURITY_CODE_ERROR,
    SecurityCodeErrorType.MfaMaxRetries
  ),
  [ERROR_CODES.VERIFY_PHONE_NUMBER_MAX_CODES_SENT]: pathWithQueryParam(
    PATH_NAMES["SECURITY_CODE_REQUEST_EXCEEDED"],
    SECURITY_CODE_ERROR,
    SecurityCodeErrorType.OtpMaxCodesSent
  ),
  [ERROR_CODES.VERIFY_PHONE_NUMBER_CODE_REQUEST_BLOCKED]: pathWithQueryParam(
    PATH_NAMES["SECURITY_CODE_WAIT"],
    SECURITY_CODE_ERROR,
    SecurityCodeErrorType.OtpBlocked
  ),
  [ERROR_CODES.ENTERED_INVALID_VERIFY_PHONE_NUMBER_CODE_MAX_TIMES]:
    pathWithQueryParam(
      PATH_NAMES["SECURITY_CODE_INVALID"],
      SECURITY_CODE_ERROR,
      SecurityCodeErrorType.OtpMaxRetries
    ),
  [ERROR_CODES.VERIFY_EMAIL_CODE_REQUEST_BLOCKED]: pathWithQueryParam(
    PATH_NAMES["SECURITY_CODE_REQUEST_EXCEEDED"],
    SECURITY_CODE_ERROR,
    SecurityCodeErrorType.EmailBlocked
  ),
  [ERROR_CODES.ENTERED_INVALID_VERIFY_EMAIL_CODE_MAX_TIMES]: pathWithQueryParam(
    PATH_NAMES["SECURITY_CODE_INVALID"],
    SECURITY_CODE_ERROR,
    SecurityCodeErrorType.EmailMaxRetries
  ),
  [ERROR_CODES.ENTERED_INVALID_PASSWORD_RESET_CODE_MAX_TIMES]:
    pathWithQueryParam(
      PATH_NAMES["SECURITY_CODE_INVALID"],
      SECURITY_CODE_ERROR,
      SecurityCodeErrorType.InvalidPasswordResetCodeMaxRetries
    ),
  [ERROR_CODES.VERIFY_EMAIL_MAX_CODES_SENT]: pathWithQueryParam(
    PATH_NAMES["SECURITY_CODE_WAIT"],
    SECURITY_CODE_ERROR,
    SecurityCodeErrorType.EmailMaxCodesSent
  ),
  [ERROR_CODES.AUTH_APP_INVALID_CODE_MAX_ATTEMPTS_REACHED]: pathWithQueryParam(
    PATH_NAMES["SECURITY_CODE_INVALID"],
    SECURITY_CODE_ERROR,
    SecurityCodeErrorType.AuthAppMfaMaxRetries
  ),
  [ERROR_CODES.VERIFY_CHANGE_HOW_GET_SECURITY_CODES_MAX_CODES_SENT]:
    pathWithQueryParam(
      PATH_NAMES["SECURITY_CODE_REQUEST_EXCEEDED"],
      SECURITY_CODE_ERROR,
      SecurityCodeErrorType.ChangeSecurityCodesEmailMaxCodesSent
    ),
  [ERROR_CODES.VERIFY_CHANGE_HOW_GET_SECURITY_CODES_INVALID_CODE]:
    pathWithQueryParam(
      PATH_NAMES["SECURITY_CODE_INVALID"],
      SECURITY_CODE_ERROR,
      SecurityCodeErrorType.ChangeSecurityCodesEmailMaxRetries
    ),
  [ERROR_CODES.VERIFY_CHANGE_HOW_GET_SECURITY_CODES_CODE_REQUEST_BLOCKED]:
    pathWithQueryParam(
      PATH_NAMES["SECURITY_CODE_WAIT"],
      SECURITY_CODE_ERROR,
      SecurityCodeErrorType.ChangeSecurityCodesEmailBlocked
    ),
};

export function getErrorPathByCode(errorCode: number): string | undefined {
  const nextPath = ERROR_CODE_MAPPING[errorCode.toString()];

  if (!nextPath) {
    return undefined;
  }

  return nextPath;
}

export function pathWithQueryParam(
  path: string,
  queryParam?: string,
  value?: string | SecurityCodeErrorType
): string {
  if (queryParam && value) {
    const queryParams = new URLSearchParams({
      [queryParam]: value,
    }).toString();

    return path + "?" + queryParams;
  }

  return path;
}

export function getNextPathAndUpdateJourney(
  req: any,
  path: string,
  event: string,
  ctx?: any,
  sessionId?: string
): string {
  const nextState = getNextState(path, event, ctx);

  req.session.user.journey = {
    nextPath: nextState.value,
    optionalPaths:
      Object.keys(nextState.meta).length > 0
        ? nextState.meta["AUTH." + nextState.value].optionalPaths
        : [],
  };

  req.log.info(
    `User journey transitioned from ${req.path} to ${nextState.value} with session id ${sessionId}`
  );

  if (!nextState) {
    throw Error(
      `Invalid user journey. No transition found from ${path} with event ${event}`
    );
  }

  return nextState.value;
}

export const JOURNEY_TYPE = {
  SIGN_IN: "sign-in",
  CREATE_ACCOUNT: "create-account",
};
