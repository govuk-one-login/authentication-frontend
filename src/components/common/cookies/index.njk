{% extends "common/layout/base.njk" %}
{% from "govuk/components/table/macro.njk" import govukTable %}
{% from "govuk/components/radios/macro.njk" import govukRadios %}
{% from "govuk/components/button/macro.njk" import govukButton %}
{% from "govuk/components/notification-banner/macro.njk" import govukNotificationBanner %}

{% set pageTitleName = 'pages.cookiePolicy.title' | translate %}

{% block content %}

{% set html %}
<h3 class="govuk-notification-banner__heading">
  {{ 'pages.cookiePolicy.successBanner.header' | translate }}
</h3>
<p class="govuk-body">{{ 'pages.cookiePolicy.successBanner.paragraph1' | translate }}</p>
<a class="govuk-notification-banner__link" id="go-back-link" href="#">{{ 'pages.cookiePolicy.successBanner.linkText' | translate }}</a>
{% endset %}

{% if updated %}
{{
  govukNotificationBanner({
  html: html,
  type: 'success',
  titleText: 'general.cookie.cookieBanner.cookieBannerTitleText' | translate,
  attributes: {
    "id": "save-success-banner"
  }
}) }}
{% else %}
{{ govukNotificationBanner({
  html: html,
  type: 'success',
  titleText: 'general.cookie.cookieBanner.cookieBannerTitleText' | translate,
  attributes: {
    "id": "save-success-banner",
    "hidden": true
  }
}) }}
{% endif %}

<h1 class="govuk-heading-l govuk-!-margin-top-0 govuk-!-margin-bottom-3">
  {{ 'pages.cookiePolicy.header' | translate }}
</h1>

<p class="govuk-body">{{'pages.cookiePolicy.info.paragraph1' | translate}}
  <a href="{{'pages.cookiePolicy.info.govUkCookiesLinkHref' | translate}}" class="govuk-link" rel="noreferrer noopener" target="_blank">{{'pages.cookiePolicy.info.govUkCookiesLinkText' | translate}}</a>{{'pages.cookiePolicy.info.paragraph2' | translate}}
</p>
<p class="govuk-body">{{'pages.cookiePolicy.info.paragraph3' | translate}}</p>
<p class="govuk-body">{{'pages.cookiePolicy.info.paragraph4' | translate}}</p>
<p class="govuk-body">{{'pages.cookiePolicy.info.paragraph5' | translate}}</p>
<p class="govuk-body">{{'pages.cookiePolicy.info.paragraph6' | translate}}
  <a href="{{'pages.cookiePolicy.info.linkHref' | translate}}" class="govuk-link" rel="noreferrer noopener" target="_blank">{{'pages.cookiePolicy.info.linkText' | translate}}</a>{{'pages.cookiePolicy.info.paragraph7' | translate}}
</p>

<h2 class="govuk-heading-s">{{'pages.cookiePolicy.essential.header' | translate}}</h2>

<p class="govuk-body">{{'pages.cookiePolicy.essential.info.paragraph1' | translate}}</p>
<ul class="govuk-list govuk-list--bullet">
  <li>{{'pages.cookiePolicy.essential.info.bulletPoint1' | translate}}</li>
  <li>{{'pages.cookiePolicy.essential.info.bulletPoint2' | translate}}</li>
  <li>{{'pages.cookiePolicy.essential.info.bulletPoint3' | translate}}</li>
</ul>

{{ govukTable({
  firstCellIsHeader: false,
  head: [
    {
    text: 'pages.cookiePolicy.essential.info.table.headers.col1' | translate,
    classes: 'govuk-body-s'
    },
    {
    text: 'pages.cookiePolicy.essential.info.table.headers.col2' | translate,
    classes: 'govuk-body-s'
    }
  ],
  rows: [
    [
      {
      text: "aps",
      classes: 'govuk-body-s'
      },
      {
      text: 'pages.cookiePolicy.essential.info.table.rows.row1.col2' | translate,
      classes: 'govuk-body-s'
      }
    ],
    [
      {
      text: "_csrf",
      classes: 'govuk-body-s'
      },
      {
      text: 'pages.cookiePolicy.essential.info.table.rows.row2.col2' | translate,
      classes: 'govuk-body-s'
      }
    ],
    [
      {
      text: "gs",
      classes: 'govuk-body-s'
      },
      {
      text: 'pages.cookiePolicy.essential.info.table.rows.row3.col2' | translate,
      classes: 'govuk-body-s'
      }
    ],
    [
      {
      text: "di-persistent-session-id",
      classes: 'govuk-body-s'
      },
      {
      text: 'pages.cookiePolicy.essential.info.table.rows.row4.col2' | translate,
      classes: 'govuk-body-s'
      }
    ],
    [
      {
      text: "am",
      classes: 'govuk-body-s'
      },
      {
      text: 'pages.cookiePolicy.essential.info.table.rows.row5.col2' | translate,
      classes: 'govuk-body-s'
      }
    ],
    [
      {
      text: "lo",
      classes: 'govuk-body-s'
      },
      {
      text: 'pages.cookiePolicy.essential.info.table.rows.row6.col2' | translate,
      classes: 'govuk-body-s'
      }
    ],
    [
      {
      text: "hmpo-wizard-sc",
      classes: 'govuk-body-s'
      },
      {
      text: 'pages.cookiePolicy.essential.info.table.rows.row7.col2' | translate,
      classes: 'govuk-body-s'
      }
    ],
    [
      {
      text: "service_session",
      classes: 'govuk-body-s'
      },
      {
      text: 'pages.cookiePolicy.essential.info.table.rows.row8.col2' | translate,
      classes: 'govuk-body-s'
      }
    ],
    [
      {
      text: "cri_passport_service_session",
      classes: 'govuk-body-s'
      },
      {
      text: 'pages.cookiePolicy.essential.info.table.rows.row9.col2' | translate,
      classes: 'govuk-body-s'
      }
    ],
    [
      {
      text: "ipv_core_service_session",
      classes: 'govuk-body-s'
      },
      {
      text: 'pages.cookiePolicy.essential.info.table.rows.row10.col2' | translate,
      classes: 'govuk-body-s'
      }
    ],
    [
      {
      text: "chl",
      classes: 'govuk-body-s'
      },
      {
      text: 'pages.cookiePolicy.essential.info.table.rows.row11.col2' | translate,
      classes: 'govuk-body-s'
      }
    ],
    [
      {
      text: "re",
      classes: 'govuk-body-s'
      },
      {
      text: 'pages.cookiePolicy.essential.info.table.rows.row12.col2' | translate,
      classes: 'govuk-body-s'
      }
    ]
  ]
}) }}

<h3 class="govuk-heading-s">{{'pages.cookiePolicy.cookiesMessage.header' | translate}}</h3>
<p class="govuk-body">{{'pages.cookiePolicy.cookiesMessage.paragraph1' | translate}}</p>
<ul class="govuk-list govuk-list--bullet">
  <li>{{'pages.cookiePolicy.cookiesMessage.bulletPoint1' | translate}}</li>
  <li>{{'pages.cookiePolicy.cookiesMessage.bulletPoint2' | translate}}</li>
</ul>

{{ govukTable({
  firstCellIsHeader: false,
  head: [
    {
      text: 'pages.cookiePolicy.cookiesMessage.table.headers.col1' | translate,
      classes: 'govuk-body-s'
    },
    {
      text: 'pages.cookiePolicy.cookiesMessage.table.headers.col2' | translate,
      classes: 'govuk-body-s'
    },
    {
      text: 'pages.cookiePolicy.cookiesMessage.table.headers.col3' | translate,
      classes: 'govuk-body-s'
    }
  ],
  rows: [
    [
    {
      text: "cookies_preferences_set",
      classes: 'govuk-body-s'
    },
    {
      text: 'pages.cookiePolicy.cookiesMessage.table.rows.row1.col2' | translate,
      classes: 'govuk-body-s'
    },
    {
      text: 'pages.cookiePolicy.cookiesMessage.table.rows.row1.col3' | translate,
      classes: 'govuk-body-s'
    }
    ]
  ]
}) }}

<h2 class="govuk-heading-s">{{'pages.cookiePolicy.settingsCookies.header' | translate}}</h2>
<p class="govuk-body">{{'pages.cookiePolicy.settingsCookies.paragraph1' | translate}}</p>
<p class="govuk-body">{{'pages.cookiePolicy.settingsCookies.paragraph2' | translate}}</p>

{{ govukTable({
  firstCellIsHeader: false,
  head: [
    {
      text: 'pages.cookiePolicy.settingsCookies.table.headers.col1' | translate,
      classes: 'govuk-body-s'
    },
    {
      text: 'pages.cookiePolicy.settingsCookies.table.headers.col2' | translate,
      classes: 'govuk-body-s'
    },
    {
      text: 'pages.cookiePolicy.settingsCookies.table.headers.col3' | translate,
      classes: 'govuk-body-s'
    }
  ],
  rows: [
    [
    {
      text: "lng",
      classes: 'govuk-body-s'
    },
    {
      text: 'pages.cookiePolicy.settingsCookies.table.rows.row1.col2' | translate,
      classes: 'govuk-body-s'
    },
    {
      text: 'pages.cookiePolicy.settingsCookies.table.rows.row1.col3' | translate,
      classes: 'govuk-body-s'
    }
    ]
  ]
}) }}

<h2 class="govuk-heading-s">{{'pages.cookiePolicy.analytics.header' | translate}}</h2>
<p class="govuk-body">{{'pages.cookiePolicy.analytics.info.paragraph1' | translate}}</p>
<p class="govuk-body">{{'pages.cookiePolicy.analytics.info.paragraph2' | translate}}</p>

{{ govukTable({
  firstCellIsHeader: false,
  head: [
    {
      text: 'pages.cookiePolicy.analytics.table.headers.col1' | translate,
      classes: 'govuk-body-s'
    },
    {
      text: 'pages.cookiePolicy.analytics.table.headers.col2' | translate,
      classes: 'govuk-body-s'
    },
    {
      text: 'pages.cookiePolicy.analytics.table.headers.col3' | translate,
      classes: 'govuk-body-s'
    }
  ],
  rows: [
    [
    {
      text: "_ga",
      classes: 'govuk-body-s'
    },
    {
      text: 'pages.cookiePolicy.analytics.table.rows.row1.col2' | translate,
      classes: 'govuk-body-s'
    },
    {
      text: 'pages.cookiePolicy.analytics.table.rows.row1.col3' | translate,
      classes: 'govuk-body-s'
    }
    ],
    [
    {
      text: "_gid",
      classes: 'govuk-body-s'
    },
    {
      text: 'pages.cookiePolicy.analytics.table.rows.row2.col2' | translate,
      classes: 'govuk-body-s'
    },
    {
      text: 'pages.cookiePolicy.analytics.table.rows.row2.col3' | translate,
      classes: 'govuk-body-s'
    }
    ],
    [
    {
      text: "_gat_UA-[number]",
      classes: 'govuk-body-s'
    },
    {
      text: 'pages.cookiePolicy.analytics.table.rows.row3.col2' | translate,
      classes: 'govuk-body-s'
    },
    {
      text: 'pages.cookiePolicy.analytics.table.rows.row3.col3' | translate,
      classes: 'govuk-body-s'
    }
    ]
  ]
}) }}

<form method="post" id="cookie-preferences-form" novalidate hidden="true">
<input type="hidden" name="_csrf" value="{{csrfToken}}"/>
{{ govukRadios({
  name: "cookie_preferences",
  attributes: {
    "id": "radio-cookie-preferences"
  },
  items: [
    {
      value: "true",
      text: 'pages.cookiePolicy.settings.yesRadioButton' | translate,
      id: "policy-cookies-accepted",
      checked: analyticsConsent === true
    },
    {
      value: "false",
      text: 'pages.cookiePolicy.settings.noRadioButton' | translate,
      id: "policy-cookies-rejected",
      checked: analyticsConsent === false
    }
  ]
}) }}

{{ govukButton({
  "text": 'pages.cookiePolicy.settings.saveButton' | translate,
  "type": "submit",
  "preventDoubleClick": true,
  attributes: {
    "id": "save-cookie-settings"
  }
}) }}

</form>

{% endblock %}
