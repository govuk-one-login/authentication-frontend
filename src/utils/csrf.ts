import { csrfSync } from "csrf-sync";

const { csrfSynchronisedProtection } = csrfSync({
  getTokenFromRequest: (req) => {
    return req.body["_csrf"];
  },
});

export { csrfSynchronisedProtection };
