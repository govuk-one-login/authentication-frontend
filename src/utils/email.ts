export function redactEmail(value: string): string {
  const asterisks = "***";
  let prefix = "";
  let domain = "";

  const email_fragments = value.split("@");

  if (email_fragments.length === 2) {
    [prefix, domain] = email_fragments;
    return `${prefix.substring(0, 3)}${asterisks}@${domain}`;
  }
  return "";
}
