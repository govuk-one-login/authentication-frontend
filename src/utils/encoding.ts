import { InvalidBase64Error } from "./error.js";
export function base64DecodeToUint8Array(value: string): Uint8Array {
  const padding = "=".repeat((4 - (value.length % 4)) % 4);
  const base64 = (value + padding).replace(/-/g, "+").replace(/_/g, "/");

  const isBase64 =
    /^([A-Za-z0-9+/]{4})*([A-Za-z0-9+/]{3}=|[A-Za-z0-9+/]{2}==)?$/.test(base64);
  if (!isBase64) {
    throw new InvalidBase64Error("String is not valid base64: " + value);
  }

  const rawData = Buffer.from(base64, "base64");
  return new Uint8Array(rawData);
}
