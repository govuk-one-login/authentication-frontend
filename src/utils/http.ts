import type {
  AxiosInstance,
  AxiosRequestConfig,
  AxiosError,
  AxiosResponse,
  AxiosRequestHeaders,
  CreateAxiosDefaults,
} from "axios";
import axios from "axios";
import {
  getApiKey,
  getFrontendApiBaseUrl,
  supportHttpKeepAlive,
} from "../config.js";
import type { ApiResponseResult } from "../types.js";
import { HTTP_STATUS_CODES } from "../app.constants.js";
import { ApiError } from "./error.js";
import type { Request } from "express";
import { createPersonalDataHeaders } from "@govuk-one-login/frontend-passthrough-headers";
import { logger } from "./logger.js";
import { Agent } from "https";

type CustomAxiosRequestHeaders = Partial<AxiosRequestHeaders>;

const headers: CustomAxiosRequestHeaders = {
  Accept: "application/json",
  "Content-Type": "application/json; charset=utf-8",
  "Access-Control-Allow-Credentials": "true",
  "X-Requested-With": "XMLHttpRequest",
};

export interface ConfigOptions {
  sessionId?: string;
  clientSessionId?: string;
  validationStatuses?: number[];
  persistentSessionId?: string;
  baseURL?: string;
  userLanguage?: string;
  reauthenticate?: boolean;
}

export function createApiResponse<T>(
  response: AxiosResponse,
  status: number[] = [HTTP_STATUS_CODES.OK]
): ApiResponseResult<T> {
  return {
    success: status.includes(response.status),
    data: response.data,
  };
}

function getSecurityHeaders(path: string, req: Request, baseUrl?: string) {
  let personalDataHeaders = {};
  let url = "";
  try {
    const basePath = baseUrl || getFrontendApiBaseUrl();
    url = new URL(path, basePath).toString();
    personalDataHeaders = createPersonalDataHeaders(url, req);
  } catch (err) {
    logger.warn(
      `Called with ${url}. Failed to set security headers due to: ${err.message}`
    );
  }
  return personalDataHeaders;
}

export function getInternalRequestConfigWithSecurityHeaders(
  options: ConfigOptions,
  req: Request,
  path: string
): AxiosRequestConfig {
  const config: AxiosRequestConfig = {
    headers: {
      "X-API-Key": getApiKey(),
      ...getSecurityHeaders(path, req, options.baseURL),
    },
    proxy: false,
  };

  if (options.sessionId) {
    config.headers["Session-Id"] = options.sessionId;
  }

  if (options.clientSessionId) {
    config.headers["Client-Session-Id"] = options.clientSessionId;
  }

  if (options.validationStatuses) {
    config.validateStatus = function (status: number) {
      return options.validationStatuses.includes(status);
    };
  }

  if (options.persistentSessionId) {
    config.headers["di-persistent-session-id"] = options.persistentSessionId;
  }

  if (options.baseURL) {
    config.baseURL = options.baseURL;
  }

  if (options.reauthenticate) {
    config.headers["Reauthenticate"] = options.reauthenticate;
  }

  if (options.userLanguage) {
    config.headers["User-Language"] = options.userLanguage;
  }

  return config;
}

export class Http {
  private instance: AxiosInstance;

  get client(): AxiosInstance {
    return this.instance || this.initHttp();
  }

  private static handleError(error: AxiosError) {
    let apiError;

    if (error.response && error.response.data) {
      apiError = new ApiError(
        error.message,
        error.response.status,
        error.response.data
      );
    } else {
      apiError = new ApiError(error.message);
    }

    return Promise.reject(apiError);
  }

  private initHttp() {
    const http = axios.create({
      baseURL: getFrontendApiBaseUrl(),
      headers: headers,
      validateStatus: (status) => {
        return (
          status >= HTTP_STATUS_CODES.OK &&
          status <= HTTP_STATUS_CODES.BAD_REQUEST
        );
      },
      ...getAdditionalAxiosConfig(),
    });

    http.interceptors.response.use(
      (response) => response,
      (error) => Http.handleError(error)
    );

    this.instance = http;
    return http;
  }
}

export function getAdditionalAxiosConfig(): CreateAxiosDefaults {
  return supportHttpKeepAlive()
    ? { httpsAgent: new Agent({ keepAlive: true }) }
    : {};
}

export const http = new Http();
