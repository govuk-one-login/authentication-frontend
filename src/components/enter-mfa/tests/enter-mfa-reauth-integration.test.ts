import { describe } from "mocha";
import { request, sinon } from "../../../../test/utils/test-utils";
import nock = require("nock");
import * as cheerio from "cheerio";
import decache from "decache";
import { AxiosResponse } from "axios";
import {
  API_ENDPOINTS,
  HTTP_STATUS_CODES,
  PATH_NAMES,
} from "../../../app.constants";
import { ERROR_CODES } from "../../common/constants";
import {
  AccountRecoveryInterface,
  AccountRecoveryResponse,
} from "../../common/account-recovery/types";
import { createApiResponse } from "../../../utils/http";
import { NextFunction, Request, Response } from "express";
import { getPermittedJourneyForPath } from "../../../../test/helpers/session-helper";

describe("Integration:: enter mfa", () => {
  let token: string | string[];
  let cookies: string;
  let app: any;
  let baseApi: string;
  const PHONE_NUMBER = "7867";
  const EXAMPLE_REDIRECT_URI = "https://rp.host/redirect";

  before(async () => {
    decache("../../../app");
    decache("../../../middleware/session-middleware");
    decache("../../common/account-recovery/account-recovery-service");
    const sessionMiddleware = require("../../../middleware/session-middleware");
    const accountRecoveryService = require("../../common/account-recovery/account-recovery-service");

    sinon
      .stub(sessionMiddleware, "validateSessionMiddleware")
      .callsFake(function (
        req: Request,
        res: Response,
        next: NextFunction
      ): void {
        res.locals.sessionId = "tDy103saszhcxbQq0-mjdzU854";
        req.session.user = {
          email: "test@test.com",
          phoneNumber: PHONE_NUMBER,
          redactedPhoneNumber: PHONE_NUMBER,
          reauthenticate: "12345",
          journey: getPermittedJourneyForPath(PATH_NAMES.ENTER_MFA),
        };
        req.session.client = {
          redirectUri: EXAMPLE_REDIRECT_URI,
        };
        next();
      });

    sinon
      .stub(accountRecoveryService, "accountRecoveryService")
      .callsFake((): AccountRecoveryInterface => {
        async function accountRecovery() {
          const fakeAxiosResponse: AxiosResponse = {
            data: {
              accountRecoveryPermitted: true,
            },
            status: HTTP_STATUS_CODES.OK,
          } as AxiosResponse;

          return createApiResponse<AccountRecoveryResponse>(fakeAxiosResponse);
        }

        return { accountRecovery };
      });

    app = await require("../../../app").createApp();
    baseApi = process.env.FRONTEND_API_BASE_URL || "";
    process.env.SUPPORT_REAUTHENTICATION = "1";

    await request(app, (test) => test.get(PATH_NAMES.ENTER_MFA), {
      expectAnalyticsPropertiesMatchSnapshot: false,
    }).then((res) => {
      const $ = cheerio.load(res.text);
      token = $("[name=_csrf]").val();
      cookies = res.headers["set-cookie"];
    });
  });

  beforeEach(() => {
    nock.cleanAll();
  });

  after(() => {
    process.env.SUPPORT_REAUTHENTICATION = "0";
    sinon.restore();
    app = undefined;
  });

  it("should return check your phone page with reauth analytics properties", async () => {
    await request(app, (test) => test.get(PATH_NAMES.ENTER_MFA).expect(200));
  });

  it("should redirect to rp if user entered 6 incorrect codes in the reauth journey", async () => {
    nock(baseApi).post(API_ENDPOINTS.VERIFY_CODE).times(6).reply(400, {
      code: ERROR_CODES.ENTERED_INVALID_MFA_MAX_TIMES,
      success: false,
    });

    await request(app, (test) =>
      test
        .post(PATH_NAMES.ENTER_MFA)
        .type("form")
        .set("Cookie", cookies)
        .send({
          _csrf: token,
          code: "123455",
        })
        .expect(
          "Location",
          EXAMPLE_REDIRECT_URI.concat("?error=login_required")
        )
        .expect(302)
    );
  });
});
