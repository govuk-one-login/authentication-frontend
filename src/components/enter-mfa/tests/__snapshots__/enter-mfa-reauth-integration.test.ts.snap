// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration:: enter mfa should return check your phone page with reauth taxonomy 1`] = `
Object {
  "taxonomyLevel1": "accounts",
  "taxonomyLevel2": "re auth",
}
`;
